sink/src/main/java/com/spredfast/kafka/connect/s3/sink/S3SinkTask.java

```diff
@@ -168,26 +162,17 @@ public void put(Collection<SinkRecord> records) throws ConnectException {
                   tp,
                   lastOffset);
 
-              PartitionWriter writer =
-                  partitions.computeIfAbsent(tp, t -> initWriter(t, firstOffset, firstTimestamp));
-              writer.writeAll(rs);
-            });
-  }
+              for (SinkRecord r : rs) {
+                PartitionWriter writer = partitions.computeIfAbsent(tp, t -> initWriter(t, r));
 
-  @Override
-  public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) throws ConnectException {
-    Metrics.StopTimer timer = metrics.time("flush", tags);
+                if (writer.shouldFlushBefore(r)) {
```

Comment:
What about the following instead?
```java
writer.writeRecord(r);
if (writer.shouldFlushAfter(r)) {
  writer.done();
  writer = partitions.computeIfAbsent(tp, t -> initWriter(t, r));
}
```

Reply:
I've added `shouldFlushAfter` method and moved the check for size based flushing there. This way if the file size threshold is exceeded the s3 file upload will happen right away. However we still need `shouldFlushBefore` since the record exceeding flushing interval should not be written to the partition writer so that it's aligned with the logic that controls flushing when there are no records for a long period of time.